### PR TITLE
[SDFAB-929] Implement builder pattern for session rules

### PR DESCRIPTION
**Pull request by EmanueleGallone**
PR's aim is to implement builder pattern to ease session's rule creation.

TODO:
 - [x] Implement FAR builder
 - [x] Implement QER builder
 - [x] Implement PDR builder

### CODE REVIEW

File: pkg/pfcpsim/session/far_builder.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package session
+
+import (
+	"github.com/wmnsk/go-pfcp/ie"
+)
+
+type farBuilder struct {
```
Comment (EmanueleGallone) [inline pkg/pfcpsim/session/far_builder.go:7]: using lowercase because I'd like to avoid exposing directly the builder struct

Comment (osinstom) [reply pkg/pfcpsim/session/far_builder.go:7]: that's fine

File: pkg/pfcpsim/session/far_builder.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	return b
+}
+
+func (b *farBuilder) WithAction(action uint8) *farBuilder {
+	b.applyAction = action
+	return b
+}
+
+func (b *farBuilder) BuildUplinkFAR() *ie.IE {
+	createFunc := ie.NewCreateFAR
```
Comment (osinstom) [inline pkg/pfcpsim/session/far_builder.go:45]: I believe we should have just a single `BuildFAR()`. The decision whether to build uplink/downlink FAR should be provided by user, e.g. by `MarkUplink()` or `MarkDownlink()`.

File: pkg/pfcpsim/session/far_builder.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/wmnsk/go-pfcp/ie"
+)
+
+type farBuilder struct {
+	farID       uint32
+	applyAction uint8
+	method      IEMethod
+}
+
+type downlinkFARBuilder struct {
```
Comment (osinstom) [inline pkg/pfcpsim/session/far_builder.go:13]: single builder instance is enough IMO (see my comment below about marking uplink/downlink)

File: pkg/pfcpsim/session/far_builder.go (hunk 2 above)
Comment (EmanueleGallone) [reply pkg/pfcpsim/session/far_builder.go:45]: Does this apply to `pdr_builder.go` as well?

Comment (osinstom) [reply pkg/pfcpsim/session/far_builder.go]: Yes

File: pkg/pfcpsim/session/far_builder.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	b.teid = teid
+	return b
+}
+
+func (b *farBuilder) WithDownlinkIP(downlinkIP string) *farBuilder {
+	b.downlinkIP = downlinkIP
+	return b
+}
+
+func (b *farBuilder) MarkAsDownlink() *farBuilder {
```
Comment (osinstom) [inline pkg/pfcpsim/session/far_builder.go:46]: I believe that we should also have `MarkAsUplink()`. We shouldn't assume uplink FAR/PDR by default and we should verify that the direction is set in `BuildFAR()`. I believe it'd improve the verbosity of API.

File: pkg/pfcpsim/session/far_builder.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	b.isDownlinkFAR = true
+	return b
+}
+
+// BuildFAR returns by default an UplinkFAR.
+// Returns a DownlinkFAR if MarkAsDownlink was invoked.
+func (b *farBuilder) BuildFAR() *ie.IE {
+	createFunc := ie.NewCreateFAR
```
Comment (osinstom) [inline pkg/pfcpsim/session/far_builder.go:54]: Is it possible to also add support for "Delete FAR IE" ?

**Review by osinstom — COMMENTED** [review]
In general, LGTM. Please address these two comments and we can merge.

File: pkg/pfcpsim/session/far_builder.go (hunk 4 above)
Comment (EmanueleGallone) [reply pkg/pfcpsim/session/far_builder.go:48]: Sure. Done

File: pkg/pfcpsim/session/far_builder.go (hunk 5 above)
Comment (EmanueleGallone) [reply pkg/pfcpsim/session/far_builder.go:73]: Added but I need to further investigate if its structure is correct.

File: pkg/pfcpsim/session/pdr_builder.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	b.n3Address = n3Address
+	return b
+}
+
+func (b *pdrBuilder) WithUEAddress(ueAddress string) *pdrBuilder {
+	b.ueAddress = ueAddress
+	return b
+}
+
+func (b *pdrBuilder) WithRulesIDs(farID uint32, sessionQERID uint32, appQERID uint32) *pdrBuilder {
```
Comment (osinstom) [inline pkg/pfcpsim/session/pdr_builder.go:64]: The PFCP spec does not differentiate between sessQERID and appQERID. I believe it's something Aether-specific.

I propose to split this method into: `WithFARID()`, `WithQERID`.  The `pdrBuilder` should store a list of QERs.

Comment (EmanueleGallone) [reply pkg/pfcpsim/session/pdr_builder.go]: Done. pdrBuilder now stores QERIDs and FARIDs as slices. I used the signatures `AddQERID(qerID uint32)` and `AddFARID(farID uint32)` to store new `ie.QERID`s `ie.FARID`s, respectively

Comment (osinstom) [reply pkg/pfcpsim/session/pdr_builder.go]: Thanks, but we should not allow to specify more than one FAR ID for a PDR, see Table 7.5.2.2-1 in the PFCP spec. It is allowed to pass several QER IDs though.

**Review by osinstom — APPROVED** [review]
LGTM!

**Pull request merged**